Modified hexdoc renderer to have a light mode and read draw settings from CSS

JohnDog3112 opened the pull request:
### Changes
* Made my hexdoc renderer read draw settings from CSS
* Added some cleanup code to delete previously generated images when switching renderers (This was already done when the page is closed/refreshed)
* Added overrides for the intersection points and segment renderer triangles when in light mode (made them black)

### Things to Note
* The rendered images don't automatically update when the CSS is changed as there was no event for it and I didn't want to have the client constantly checking for changes.
* Since you need an html element with the CSS styles to get their values, the only thing used to check for changes is if you selected a different palette or renderer.
  *  This means that to see changes from the CSS, you need to switch palettes/renderers

object-Object commented:
For the auto-update thing, I have a couple of ideas:
- Regenerate the image if you reselect the type/palette that are already selected
- Generate the image each time the pattern drawer is opened, instead of caching it

object-Object reviewed (commented):
Other than those two comments, I think it looks good, though I'm not really that familiar with frontend stuff like this.

JohnDog3112 commented:
> Also noticing some weird pixelation on the edges of the lines here - is there an antialiasing setting we could enable or something?
> 
> ![image](https://private-user-images.githubusercontent.com/37044997/328973760-c81336ba-8fc4-4dec-8b6b-25593e58272b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyODkyNDksIm5iZiI6MTcxNTI4ODk0OSwicGF0aCI6Ii8zNzA0NDk5Ny8zMjg5NzM3NjAtYzgxMzM2YmEtOGZjNC00ZGVjLThiNmItMjU1OTNlNTgyNzJiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA5VDIxMDkwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkNTg5MDA2MDdiNzRkNzY1MzViYjk1NjdkZDZjNmYzNWRlYjY2ZTg0ZWJhYmMxMTA2MGQ4YWYyZmFmY2VkNTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.AQDOyMKREtspJURkwOW3ZO0eI0iZnLEZ8lX_rWeYSa4)

So, it turns out, anti-aliasing is enabled for every pattern except gradient... I'm not quite sure how that happened but it's a pretty easy fix.

JohnDog3112 commented:
> For the auto-update thing, I have a couple of ideas:
> 
>     * Regenerate the image if you reselect the type/palette that are already selected
> 
>     * Generate the image each time the pattern drawer is opened, instead of caching it

I could do either? Though, for both, I would probably add a function to check for changes in the css when reselecting or opening a drawer. I also thought of starting a timer when it detects a manual change in the css (after doing one of the above things)

JohnDog3112 commented:
> > Also noticing some weird pixelation on the edges of the lines here - is there an antialiasing setting we could enable or something?
> > ![image](https://private-user-images.githubusercontent.com/37044997/328973760-c81336ba-8fc4-4dec-8b6b-25593e58272b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyODkyNDksIm5iZiI6MTcxNTI4ODk0OSwicGF0aCI6Ii8zNzA0NDk5Ny8zMjg5NzM3NjAtYzgxMzM2YmEtOGZjNC00ZGVjLThiNmItMjU1OTNlNTgyNzJiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA5VDIxMDkwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkNTg5MDA2MDdiNzRkNzY1MzViYjk1NjdkZDZjNmYzNWRlYjY2ZTg0ZWJhYmMxMTA2MGQ4YWYyZmFmY2VkNTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.AQDOyMKREtspJURkwOW3ZO0eI0iZnLEZ8lX_rWeYSa4)
> 
> So, it turns out, anti-aliasing is enabled for every pattern except gradient... I'm not quite sure how that happened but it's a pretty easy fix.

Fixed 
![image](https://github.com/FallingColors/HexMod/assets/16504051/5a8c863d-164b-4cba-a16b-7b688017268c)

JohnDog3112 commented:
I implemented two things for updating from css.
1. Made it so selecting same option refreshes it
2. Added a function (autoUpdateRenderer) that sets up a function to run every x ms and re-renders the images if it detects a change in the settings. It only checks the css of the first image it finds, so it won't detect changes in individual image css values, but I'm not sure why someone would do that rather than use a css class.

object-Object commented on doc/src/hexdoc_hexcasting/_templates/hexcasting_render.js.jinja:
```diff
@@ -319,19 +319,80 @@ let options = {
+  selected = name;
+}
+
+function checkEqual(ob1, ob2) {
```
huh, none of the libraries we're using have a thing for this?

object-Object replied:
hexdoc uses jquery and bootstrap:

https://github.com/hexdoc-dev/hexdoc/blob/b76ee52d09a1f5e829af0eb2b2f80629806a8569/src/hexdoc/_templates/index.html.jinja#L26-L38

JohnDog3112 replied:
It doesn't seem like either includes a function like that.

object-Object commented on doc/src/hexdoc_hexcasting/_templates/hexcasting_render.js.jinja:
```diff
@@ -319,19 +319,80 @@ let options = {
+    clearInterval(autoUpdateInterval);
+  }
+
+  if (real_interval == null) {
+    real_interval = 500;
+  } else if (real_interval <= 0) {
+    return;
+  }
```
this feels confusing; why is real_interval separate from interval?

```js
interval ??= 500;
if (interval < 0) {
  console.log(`Invalid interval for autoUpdateRenderer: ${interval}`);
  return;
}
```

JohnDog3112 replied:
I think I was in a Rust mindset and forgot that I could mutate the parameter. I replaced it with a default parameter instead

object-Object commented on doc/src/hexdoc_hexcasting/_templates/hexcasting_render.js.jinja:
```diff
@@ -319,19 +319,80 @@ let options = {
+
+    let new_settings = options[selected](styles, last_palette);
+    if (!checkEqual(new_settings,old_settings)) {
+      console.log("update!");
```
slightly more detailed message would be nice :)

JohnDog3112 commented on doc/src/hexdoc_hexcasting/_templates/hexcasting_render.js.jinja:
```diff
@@ -319,19 +319,80 @@ let options = {
 
-  if (name == "animated") {
+  if (name == "animated" && selected != name) {
     load_animated();
```
selected != name needed to be nested, otherwise, it tries to run the other renderer on the second selection.

JohnDog3112 commented on doc/src/hexdoc_hexcasting/_templates/hexcasting_render.js.jinja:
```diff
@@ -319,19 +319,80 @@ let options = {
+    if (selected == "animated") {
+      return;
+    }
+    let first_pattern = document.querySelector("img.spell-viz");
```
I think so, I just have to add a check to clear it when switching to the animated renderer.
